File: typescript/packages/core/src/core/langfuse/readdir.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { LangfuseAccessor } from '../../accessor/langfuse.ts'
import { IndexEntry } from '../../cache/index/config.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import type { PathSpec } from '../../types.ts'
import {
  fetchDatasetRuns,
  fetchDatasets,
  fetchPrompts,
  fetchSessions,
  fetchTraces,
} from './_client.ts'
import { stripSlash } from '../../util/slash.ts'

const TOP_LEVEL_DIRS = ['traces', 'sessions', 'prompts', 'datasets'] as const

const DEFAULT_TRACE_LIMIT = 300
const DEFAULT_TRACE_WINDOW_MS = 7 * 24 * 60 * 60 * 1000

function defaultFromTimestamp(): string {
  return new Date(Date.now() - DEFAULT_TRACE_WINDOW_MS).toISOString()
}

function enoent(path: string): Error {
  const err = new Error(`ENOENT: ${path}`) as Error & { code: string }
  err.code = 'ENOENT'
  return err
}

function pickString(record: Record<string, unknown>, key: string): string {
  const value = record[key]
  return typeof value === 'string' ? value : ''
}

function pickStringOrNumber(record: Record<string, unknown>, key: string): string {
  const value = record[key]
  if (typeof value === 'string') return value
  if (typeof value === 'number') return String(value)
  return ''
}

function makeVirtualKey(prefix: string, key: string): string {
  if (key === '') return prefix !== '' ? prefix : '/'
  return `${prefix}/${key}`
}

async function readdirTraces(
  accessor: LangfuseAccessor,
  virtualKey: string,
  index: IndexCacheStore | undefined,
  prefix: string,
): Promise<string[]> {
  if (index !== undefined) {
    const listing = await index.listDir(virtualKey)
    if (listing.entries !== undefined && listing.entries !== null) return listing.entries
  }
  const limit = accessor.config.defaultTraceLimit ?? DEFAULT_TRACE_LIMIT
  const fromTimestamp = accessor.config.defaultFromTimestamp ?? defaultFromTimestamp()
  const traces = await fetchTraces(accessor.transport, { limit, fromTimestamp })
  const entries: [string, IndexEntry][] = []
  const names: string[] = []
  for (const t of traces) {
    const traceId = pickString(t, 'id')
    const filename = `${traceId}.json`
    entries.push([
      filename,
      new IndexEntry({
        id: traceId,
        name: traceId,
        resourceType: 'langfuse/trace',
        vfsName: filename,
      }),
    ])
    names.push(`${prefix}/traces/${filename}`)
  }
  if (index !== undefined) await index.setDir(virtualKey, entries)
  return names
}

async function readdirSessions(
  accessor: LangfuseAccessor,
  virtualKey: string,
  index: IndexCacheStore | undefined,
  prefix: string,
): Promise<string[]> {
  if (index !== undefined) {
    const listing = await index.listDir(virtualKey)
    if (listing.entries !== undefined && listing.entries !== null) return listing.entries
  }
  const sessions = await fetchSessions(accessor.transport)
  const entries: [string, IndexEntry][] = []
  const names: string[] = []
  for (const s of sessions) {
    const sessionId = pickString(s, 'id')
    entries.push([
      sessionId,
      new IndexEntry({
        id: sessionId,
        name: sessionId,
        resourceType: 'langfuse/session',
        vfsName: sessionId,
      }),
    ])
    names.push(`${prefix}/sessions/${sessionId}`)
  }
  if (index !== undefined) await index.setDir(virtualKey, entries)
  return names
}

async function readdirSessionTraces(
  accessor: LangfuseAccessor,
  sessionId: string,
  virtualKey: string,
  index: IndexCacheStore | undefined,
  prefix: string,
): Promise<string[]> {
  if (index !== undefined) {
    const listing = await index.listDir(virtualKey)
    if (listing.entries !== undefined && listing.entries !== null) return listing.entries
  }
  const limit = accessor.config.defaultTraceLimit ?? DEFAULT_TRACE_LIMIT
  const fromTimestamp = accessor.config.defaultFromTimestamp ?? defaultFromTimestamp()
  const traces = await fetchTraces(accessor.transport, { sessionId, limit, fromTimestamp })
  const entries: [string, IndexEntry][] = []
  const names: string[] = []
  for (const t of traces) {
    const traceId = pickString(t, 'id')
    const filename = `${traceId}.json`
    entries.push([
      filename,
      new IndexEntry({
        id: traceId,
        name: traceId,
        resourceType: 'langfuse/trace',
        vfsName: filename,
      }),
    ])
    names.push(`${prefix}/sessions/${sessionId}/${filename}`)
  }
  if (index !== undefined) await index.setDir(virtualKey, entries)
  return names
}

async function readdirPrompts(
  accessor: LangfuseAccessor,
  virtualKey: string,
  index: IndexCacheStore | undefined,
  prefix: string,
): Promise<string[]> {
  if (index !== undefined) {
    const listing = await index.listDir(virtualKey)
    if (listing.entries !== undefined && listing.entries !== null) return listing.entries
  }
  const prompts = await fetchPrompts(accessor.transport)
  const seen = new Set<string>()
  const entries: [string, IndexEntry][] = []
  const names: string[] = []
  for (const p of prompts) {
    const promptName = pickString(p, 'name')
    if (seen.has(promptName)) continue
    seen.add(promptName)
    entries.push([
      promptName,
      new IndexEntry({
        id: promptName,
        name: promptName,
        resourceType: 'langfuse/prompt',
        vfsName: promptName,
      }),
    ])
    names.push(`${prefix}/prompts/${promptName}`)
  }
  if (index !== undefined) await index.setDir(virtualKey, entries)
  return names
}

async function readdirPromptVersions(
  accessor: LangfuseAccessor,
  promptName: string,
  virtualKey: string,
  index: IndexCacheStore | undefined,
  prefix: string,
): Promise<string[]> {
  if (index !== undefined) {
    const listing = await index.listDir(virtualKey)
    if (listing.entries !== undefined && listing.entries !== null) return listing.entries
  }
  const prompts = await fetchPrompts(accessor.transport)
  const entries: [string, IndexEntry][] = []
  const names: string[] = []
  for (const p of prompts) {
    if (pickString(p, 'name') !== promptName) continue
    const version = pickStringOrNumber(p, 'version') || '0'
    const filename = `${version}.json`
    entries.push([
      filename,
      new IndexEntry({
        id: `${promptName}/${version}`,
        name: version,
        resourceType: 'langfuse/prompt_version',
        vfsName: filename,
      }),
    ])
    names.push(`${prefix}/prompts/${promptName}/${filename}`)
  }
  if (index !== undefined) await index.setDir(virtualKey, entries)
  return names
}

async function readdirDatasets(
  accessor: LangfuseAccessor,
  virtualKey: string,
  index: IndexCacheStore | undefined,
  prefix: string,
): Promise<string[]> {
  if (index !== undefined) {
    const listing = await index.listDir(virtualKey)
    if (listing.entries !== undefined && listing.entries !== null) return listing.entries
  }
  const datasets = await fetchDatasets(accessor.transport)
  const entries: [string, IndexEntry][] = []
  const names: string[] = []
  for (const d of datasets) {
    const datasetName = pickString(d, 'name')
    entries.push([
      datasetName,
      new IndexEntry({
        id: datasetName,
        name: datasetName,
        resourceType: 'langfuse/dataset',
        vfsName: datasetName,
      }),
    ])
    names.push(`${prefix}/datasets/${datasetName}`)
  }
  if (index !== undefined) await index.setDir(virtualKey, entries)
  return names
}

async function readdirDatasetRuns(
  accessor: LangfuseAccessor,
  datasetName: string,
  virtualKey: string,
  index: IndexCacheStore | undefined,
  prefix: string,
): Promise<string[]> {
  if (index !== undefined) {
    const listing = await index.listDir(virtualKey)
    if (listing.entries !== undefined && listing.entries !== null) return listing.entries
  }
  const runs = await fetchDatasetRuns(accessor.transport, datasetName)
  const entries: [string, IndexEntry][] = []
  const names: string[] = []
  for (const r of runs) {
    const runName = pickString(r, 'name')
    const filename = `${runName}.jsonl`
    entries.push([
      filename,
      new IndexEntry({
        id: runName,
        name: runName,
        resourceType: 'langfuse/dataset_run',
        vfsName: filename,
      }),
    ])
    names.push(`${prefix}/datasets/${datasetName}/runs/${filename}`)
  }
  if (index !== undefined) await index.setDir(virtualKey, entries)
  return names
}

export async function readdir(
  accessor: LangfuseAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<string[]> {
  const prefix = path.prefix
  let p = path.pattern !== null ? path.directory : path.original
  if (prefix !== '' && p.startsWith(prefix)) {
    p = p.slice(prefix.length) || '/'
  }
  const key = stripSlash(p)
  for (const part of key.split('/')) {
    if (key !== '' && part.startsWith('.')) throw enoent(p)
  }
  const virtualKey = makeVirtualKey(prefix, key)

  if (key === '') {
    return TOP_LEVEL_DIRS.map((d) => `${prefix}/${d}`)
  }

  const parts = key.split('/')

  if (parts[0] === 'traces' && parts.length === 1) {
    return readdirTraces(accessor, virtualKey, index, prefix)
  }

  if (parts[0] === 'sessions' && parts.length === 1) {
    return readdirSessions(accessor, virtualKey, index, prefix)
  }

  if (parts[0] === 'sessions' && parts.length === 2) {
    return readdirSessionTraces(accessor, parts[1] ?? '', virtualKey, index, prefix)
  }

  if (parts[0] === 'prompts' && parts.length === 1) {
    return readdirPrompts(accessor, virtualKey, index, prefix)
  }

  if (parts[0] === 'prompts' && parts.length === 2) {
    return readdirPromptVersions(accessor, parts[1] ?? '', virtualKey, index, prefix)
  }

  if (parts[0] === 'datasets' && parts.length === 1) {
    return readdirDatasets(accessor, virtualKey, index, prefix)
  }

  if (parts[0] === 'datasets' && parts.length === 2) {
    return [
      `${prefix}/datasets/${parts[1] ?? ''}/items.jsonl`,
      `${prefix}/datasets/${parts[1] ?? ''}/runs`,
    ]
  }

  if (parts[0] === 'datasets' && parts.length === 3 && parts[2] === 'runs') {
    return readdirDatasetRuns(accessor, parts[1] ?? '', virtualKey, index, prefix)
  }

  throw enoent(p)
}

File: python/mirage/core/langfuse/readdir.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.langfuse import LangfuseAccessor
from mirage.cache.index import IndexCacheStore, IndexEntry
from mirage.core.langfuse._client import (fetch_dataset_runs, fetch_datasets,
                                          fetch_prompts, fetch_sessions,
                                          fetch_traces)
from mirage.types import PathSpec
from mirage.utils.errors import enoent

TOP_LEVEL_DIRS = ["traces", "sessions", "prompts", "datasets"]


async def readdir(
    accessor: LangfuseAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> list[str]:
    """List directory contents.

    Args:
        accessor (LangfuseAccessor): langfuse accessor.
        path (PathSpec | str): resource-relative path.
        index (IndexCacheStore | None): index cache.
        prefix (str): mount prefix for virtual index keys.
    """
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    virtual = path.original
    if isinstance(path, PathSpec):
        prefix = path.prefix
        path = path.directory if path.pattern else path.original
    if prefix and path.startswith(prefix):
        rest = path[len(prefix):]
        if prefix.endswith("/") or rest == "" or rest.startswith("/"):
            path = rest or "/"
    key = path.strip("/")

    if key and any(p.startswith(".") for p in key.split("/")):
        raise enoent(virtual)

    virtual_key = prefix + "/" + key if key else prefix or "/"

    if not key:
        return [f"{prefix}/{d}" for d in TOP_LEVEL_DIRS]

    parts = key.split("/")

    if parts[0] == "traces" and len(parts) == 1:
        return await _readdir_traces(accessor, virtual_key, index, prefix)

    if parts[0] == "sessions" and len(parts) == 1:
        return await _readdir_sessions(accessor, virtual_key, index, prefix)

    if parts[0] == "sessions" and len(parts) == 2:
        return await _readdir_session_traces(
            accessor,
            parts[1],
            virtual_key,
            index,
            prefix,
        )

    if parts[0] == "prompts" and len(parts) == 1:
        return await _readdir_prompts(accessor, virtual_key, index, prefix)

    if parts[0] == "prompts" and len(parts) == 2:
        return await _readdir_prompt_versions(
            accessor,
            parts[1],
            virtual_key,
            index,
            prefix,
        )

    if parts[0] == "datasets" and len(parts) == 1:
        return await _readdir_datasets(accessor, virtual_key, index, prefix)

    if parts[0] == "datasets" and len(parts) == 2:
        return [
            f"{prefix}/datasets/{parts[1]}/items.jsonl",
            f"{prefix}/datasets/{parts[1]}/runs",
        ]

    if (parts[0] == "datasets" and len(parts) == 3 and parts[2] == "runs"):
        return await _readdir_dataset_runs(
            accessor,
            parts[1],
            virtual_key,
            index,
            prefix,
        )

    raise enoent(virtual)


async def _readdir_traces(
    accessor: LangfuseAccessor,
    virtual_key: str,
    index: IndexCacheStore | None,
    prefix: str,
) -> list[str]:
    if index is not None:
        listing = await index.list_dir(virtual_key)
        if listing.entries is not None:
            return listing.entries
    limit = accessor.config.default_trace_limit
    traces = await fetch_traces(accessor.api, limit=limit)
    entries = []
    names = []
    for t in traces:
        trace_id = t.get("id", "")
        filename = f"{trace_id}.json"
        entry = IndexEntry(
            id=trace_id,
            name=trace_id,
            resource_type="langfuse/trace",
            vfs_name=filename,
        )
        entries.append((filename, entry))
        names.append(f"{prefix}/traces/{filename}")
    if index is not None:
        await index.set_dir(virtual_key, entries)
    return names


async def _readdir_sessions(
    accessor: LangfuseAccessor,
    virtual_key: str,
    index: IndexCacheStore | None,
    prefix: str,
) -> list[str]:
    if index is not None:
        listing = await index.list_dir(virtual_key)
        if listing.entries is not None:
            return listing.entries
    sessions = await fetch_sessions(accessor.api)
    entries = []
    names = []
    for s in sessions:
        session_id = s.get("id", "")
        entry = IndexEntry(
            id=session_id,
            name=session_id,
            resource_type="langfuse/session",
            vfs_name=session_id,
        )
        entries.append((session_id, entry))
        names.append(f"{prefix}/sessions/{session_id}")
    if index is not None:
        await index.set_dir(virtual_key, entries)
    return names


async def _readdir_session_traces(
    accessor: LangfuseAccessor,
    session_id: str,
    virtual_key: str,
    index: IndexCacheStore | None,
    prefix: str,
) -> list[str]:
    if index is not None:
        listing = await index.list_dir(virtual_key)
        if listing.entries is not None:
            return listing.entries
    limit = accessor.config.default_trace_limit
    traces = await fetch_traces(
        accessor.api,
        session_id=session_id,
        limit=limit,
    )
    entries = []
    names = []
    for t in traces:
        trace_id = t.get("id", "")
        filename = f"{trace_id}.json"
        entry = IndexEntry(
            id=trace_id,
            name=trace_id,
            resource_type="langfuse/trace",
            vfs_name=filename,
        )
        entries.append((filename, entry))
        names.append(f"{prefix}/sessions/{session_id}/{filename}")
    if index is not None:
        await index.set_dir(virtual_key, entries)
    return names


async def _readdir_prompts(
    accessor: LangfuseAccessor,
    virtual_key: str,
    index: IndexCacheStore | None,
    prefix: str,
) -> list[str]:
    if index is not None:
        listing = await index.list_dir(virtual_key)
        if listing.entries is not None:
            return listing.entries
    prompts = await fetch_prompts(accessor.api)
    seen: set[str] = set()
    entries = []
    names = []
    for p in prompts:
        prompt_name = p.get("name", "")
        if prompt_name in seen:
            continue
        seen.add(prompt_name)
        entry = IndexEntry(
            id=prompt_name,
            name=prompt_name,
            resource_type="langfuse/prompt",
            vfs_name=prompt_name,
        )
        entries.append((prompt_name, entry))
        names.append(f"{prefix}/prompts/{prompt_name}")
    if index is not None:
        await index.set_dir(virtual_key, entries)
    return names


async def _readdir_prompt_versions(
    accessor: LangfuseAccessor,
    prompt_name: str,
    virtual_key: str,
    index: IndexCacheStore | None,
    prefix: str,
) -> list[str]:
    if index is not None:
        listing = await index.list_dir(virtual_key)
        if listing.entries is not None:
            return listing.entries
    prompts = await fetch_prompts(accessor.api)
    entries = []
    names = []
    for p in prompts:
        if p.get("name") != prompt_name:
            continue
        version = p.get("version", 0)
        filename = f"{version}.json"
        entry = IndexEntry(
            id=f"{prompt_name}/{version}",
            name=str(version),
            resource_type="langfuse/prompt_version",
            vfs_name=filename,
        )
        entries.append((filename, entry))
        names.append(f"{prefix}/prompts/{prompt_name}/{filename}")
    if index is not None:
        await index.set_dir(virtual_key, entries)
    return names


async def _readdir_datasets(
    accessor: LangfuseAccessor,
    virtual_key: str,
    index: IndexCacheStore | None,
    prefix: str,
) -> list[str]:
    if index is not None:
        listing = await index.list_dir(virtual_key)
        if listing.entries is not None:
            return listing.entries
    datasets = await fetch_datasets(accessor.api)
    entries = []
    names = []
    for d in datasets:
        dataset_name = d.get("name", "")
        entry = IndexEntry(
            id=dataset_name,
            name=dataset_name,
            resource_type="langfuse/dataset",
            vfs_name=dataset_name,
        )
        entries.append((dataset_name, entry))
        names.append(f"{prefix}/datasets/{dataset_name}")
    if index is not None:
        await index.set_dir(virtual_key, entries)
    return names


async def _readdir_dataset_runs(
    accessor: LangfuseAccessor,
    dataset_name: str,
    virtual_key: str,
    index: IndexCacheStore | None,
    prefix: str,
) -> list[str]:
    if index is not None:
        listing = await index.list_dir(virtual_key)
        if listing.entries is not None:
            return listing.entries
    runs = await fetch_dataset_runs(accessor.api, dataset_name)
    entries = []
    names = []
    for r in runs:
        run_name = r.get("name", "")
        filename = f"{run_name}.jsonl"
        entry = IndexEntry(
            id=run_name,
            name=run_name,
            resource_type="langfuse/dataset_run",
            vfs_name=filename,
        )
        entries.append((filename, entry))
        names.append(f"{prefix}/datasets/{dataset_name}/runs/{filename}")
    if index is not None:
        await index.set_dir(virtual_key, entries)
    return names
